codec/src/main/java/io/netty/handler/codec/BatchFrameDecoder.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.codec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+import java.util.List;
+
+public abstract class BatchFrameDecoder extends ChannelInboundHandlerAdapter {
+
+    private final int startBufferSize;
+
+    private ByteBuf buffer;
+
+    /**
+     * @param startBufferSize the start buffer size which is used for batching incoming {@link ByteBuf} messages. Size
+     * of this buffer must be greater or equal to zero. If the size of this buffer is zero then the default size will be
+     * used (= 256 bytes).
+     */
+    protected BatchFrameDecoder(int startBufferSize) {
+        if (startBufferSize < 0) {
+            throw new IllegalArgumentException("startBufferSize must be greater or equal to zero: " + startBufferSize);
+        }
+
+        this.startBufferSize = startBufferSize;
+
+        CodecUtil.ensureNotSharable(this);
+    }
+
+    protected abstract void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out);
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        buffer = internalBuffer(ctx, startBufferSize);
+
+        ctx.fireChannelActive();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        releaseInternalBuffer();
+        ctx.fireChannelInactive();
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        releaseInternalBuffer();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (msg instanceof ByteBuf) {
+            final ByteBuf frame = (ByteBuf) msg;
+
+            try {
+                buffer.writeBytes(frame);
```

Comment:
I think we should strive to share/extend from `ByteToMessageDecoder` as it already also covers some boundary cases for use (channel input closed, auto read, etc..), and `ByteToMessageDecoder` is widely used so if more issues arise they will likely be fixed there.

One potential approach with minimal changes to `ByteToMessageDecoder` would be to have have a class that extends `ByteToMessageDecoder` (something like `BatchedByteToMessageDecoder`) like the following:

```java
// This is just to sketch the general ideas ... this is not necessarily production ready
public class BatchedByteToMessageDecoder extends ByteToMessageDecoder {
  private boolean readInProgress;

  @Override
  public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
    readInProgress = true;
    super.channelRead(ctx, msg);
  }

  // This method may be able to be final, but if the user is not using autoRead then they may want to
  // read more data manually here.
  @Override
  protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
    if (!readInProgress) {
      // Some exception situation occurred ... dispatch this read!
      decodeBatched(ctx, in, out);
    }
    // If we are in channel read then we don't want to call user code
    // until channelReadComplete so this is a NOOP
  }

  protected abstract void decodeBatched(ChannelHandlerContext ctx, ByteBuf in, List<Object> out);

  @Override
  public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
    readInProgress = false;
    CodecOutputList out = CodecOutputList.newInstance();
    try {
      decodeBatched(ctx, internalBuffer(), out);
    } finally {
       afterDecode(out);
    }
    super.channelReadComplete(ctx);
  }
}

// this Method would be added to `ByteToMessageDecoder` and is the extracted pieces in the
// finally block of `channelRead`.
  final void afterDecode(CodecOutputList out) {
                if (cumulation != null && !cumulation.isReadable()) {
                    numReads = 0;
                    cumulation.release();
                    cumulation = null;
                } else if (++ numReads >= discardAfterReads) {
                    // We did enough reads already try to discard some bytes so we not risk to see a OOME.
                    // See https://github.com/netty/netty/issues/4275
                    numReads = 0;
                    discardSomeReadBytes();
                }

                int size = out.size();
                decodeWasNull = !out.insertSinceRecycled();
                fireChannelRead(ctx, out, size);
                out.recycle();
  }
```